release/nightly_tests/dataset/multi_node_train_benchmark.py

```diff
@@ -98,81 +146,212 @@ def get_transform(to_torch_tensor):
+            args.data_root,
+        )
     else:
         raise Exception(f"Unknown file type {args.file_type}")
-    # 2) Preprocess data by applying transformation with map_batches()
-    ray_dataset = ray_dataset.map_batches(crop_and_flip_image_batch)
+
+    if cache_input_ds:
+        ray_dataset = ray_dataset.materialize()
+
+    # 2) Preprocess data by applying transformation with map/map_batches()
+    ray_dataset = ray_dataset.map(crop_and_flip_image)
+    if cache_output_ds:
+        ray_dataset = ray_dataset.materialize()
 
     def train_loop_per_worker():
         it = train.get_dataset_shard("train")
+        device = train.torch.get_device()
 
         for i in range(args.num_epochs):
+            print(f"Epoch {i+1} of {args.num_epochs}")
             num_rows = 0
             start_t = time.time()
-            for batch in it.iter_batches(
+            for batch in it.iter_torch_batches(
                 batch_size=args.batch_size,
-                local_shuffle_buffer_size=args.local_shuffle_buffer_size,
-                prefetch_batches=10,
             ):
                 num_rows += args.batch_size
             end_t = time.time()
-            # Record throughput per epoch.
-            epoch_tput = num_rows / (end_t - start_t)
-            train.report({"tput": epoch_tput, "epoch": i})
+        # Workaround to report the final epoch time from each worker, so that we
+        # can sum up the times at the end when calculating throughput.
+        world_size = ray.train.get_context().get_world_size()
+        all_workers_time_list = [
+            torch.zeros((2), dtype=torch.double, device=device)
+            for _ in range(world_size)
+        ]
+        curr_worker_time = torch.tensor(
+            [start_t, end_t], dtype=torch.double, device=device
+        )
+        dist.all_gather(all_workers_time_list, curr_worker_time)
+
+        all_num_rows = [
+            torch.zeros((1), dtype=torch.int32, device=device)
+            for _ in range(world_size)
+        ]
+        curr_num_rows = torch.tensor([num_rows], dtype=torch.int32, device=device)
+        dist.all_gather(all_num_rows, curr_num_rows)
+
+        train.report(
+            {
+                "time_final_epoch": [
+                    tensor.tolist() for tensor in all_workers_time_list
+                ],
+                "num_rows": [tensor.item() for tensor in all_num_rows],
+            }
+        )
 
     # 3) Train TorchTrainer on processed data
     options = DataConfig.default_ingest_options()
     options.preserve_order = args.preserve_order
 
-    torch_trainer = TorchTrainer(
-        train_loop_per_worker,
-        datasets={"train": ray_dataset},
-        scaling_config=ScalingConfig(num_workers=args.num_workers, use_gpu=True),
-        dataset_config=ray.train.DataConfig(
-            execution_options=options,
-        ),
-    )
+    if prepartition_ds:
+
+        class PrepartitionCacheDataConfig(DataConfig):
```

Comment:
Shall we just remove this as it's not used?

Reply:
yeah good point, we can re-implement this correctly in a future PR.